[[ Perf ]] Optimization of native string processing

runrevmark opened the pull request:
This patch optimizes string processing for native strings in a number of ways.

It contains a re-implemented stack of low-level native string operations which take full advantage of the properties of the native encoding.

It adds a libfoundation function for 'delimited' chunk offset operations and specializations for native strings with 1 char delimiters (the most common case). These are now used by lineOffset and itemOffset.

It adds a libfoundation function for 'delimited' chunk range operations. These are now used for evaluation of 'item x [ to y ] of' and 'line x [ to y ] of'.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/livecode/livecode/3267)

<!-- Reviewable:end -->

runrevmark committed:
[[ Foundation ]] Rework native string operations.

The low-level 'native char' operations have been replaced with more
efficient variants which self-select core loops based on the current
string options and content of the strings.

runrevmark committed:
[[ ChunkOffset ]] Add specialized 'chunkOffset' implementation.

An optimized operation MCStringDelimitedOffset has been added which
performs the main process used by lineOffset and itemOffset. In
addition, a specialized form has been added for the most common
case - native haystack, native needle, native 1 character delimiter.

runrevmark committed:
[[ Chunks ]] Optimize delimited chunk evaluation.

An optimized operation MCStringForwardDelimitedRegion has been
added which implements the operation of getting a single chunk
or chunk range from a string when the chunk is of delimited type.

This commit uses the operation for the 'item' chunk where both
start and finish indicies are non-negative.

peter-b commented:
Well this looks really really nice. :grinning: I'm really looking forward to seeing the finished version!

peter-b commented:
Possible issues: this patch will currently only work with GCC >= 4.7 and C++11, because in earlier versions of C++ function arguments to templates are required to have external linkage, i.e. non-`static`.

runrevmark committed:
[[ Foundation ]] Add missing range constants for foundation types.

Minimum and maximum range constants have been added for byte_t,
codepoint_t and unichar_t.

runrevmark committed:
[[ Foundation ]] Replace 'magic numbers' when checking codepoints.

Instances of checking for a codepoint being outside the BMP have
been changed to use UNICHAR_MAX, rather than 0xFFFF.

runrevmark committed:
[[ Foundation ]] Correct a few minor issues.

Correct typos and poor variable naming in the new native char ops
code.

runrevmark committed:
Merge branch 'develop' into perf-faster_native_strings

runrevmark committed:
[[ Chunks ]] Optimized delimited chunk evaluation.

This commit makes the line chunk use MCStringForwardDelimitedRegion.

runrevmark committed:
[[ ChunkOffset ]] Fix 'wholeMatches' boundary check.

After and before found delimiters were mixed up - this has been
corrected.

runrevmark committed:
[[ Chunks ]] Fix ordinal chunks.

The ordinal chunks were incorrectly calculating the index. This
has been corrected.

runrevmark committed:
[[ ChunkOffset ]] Empty needle case broken.

The MCStringDelimitedOffset API was not checking for an empty
needle.

runrevmark committed:
[[ Chunks ]] Fix chunks of chunks.

Chained delimited chunk evaluation was not working correctly. Partly
due to incorrectly checking for errors in mark_for_eval, and partly
because of incorrect range computations in DelimitedRange.

runrevmark committed:
[[ Foundation ]] Change the MCStringForwardDelimitedRange signature.

The signature of MCStringForwardDelimitedRange has been changed to use
an MCRange for the region. Making it more consistent with the 0-based
offset and length approach to MCString functions.

runrevmark committed:
[[ Foundation ]] Minor formatting cleanups of new functions.

livecodeali commented:
This is a reminder to find and replace occurrences of 'occurrance' with 'occurrence' :-)

peter-b commented on engine/src/chunk.cpp:
```diff
@@ -2283,6 +2283,191 @@ static void skip_word(const char *&sptr, const char *&eptr)
+    if (byte != nil &&
+        !__MCCRefMarkChunkForEval(CT_BYTE, ctxt, byte, x_mark))
+        return;
+}
```
Please add an fallback (e.g. `MCUnreachable()`) for if execution falls through all these tests.

runrevmark replied:
There is no fallback here... It just processes each potential chunk type in turn - if there are none, the range doesn't get restricted.

runrevmark committed:
[[ Foundation ]] Make sure static functions have __MC prefix.

Updates all __NativeOp, __NativeStr and __NativeChar functions to
have __MC as a prefix.

runrevmark committed:
[[ Foundation ]] Correct mispellings of 'occurrence' and friends.

runrevmark committed:
[[ Foundation ]] Make sure template parameters have external linkage.

runrevmark committed:
[[ Chunks ]] Make sure template parameters have external linkage.

runrevmark committed:
[[ Foundation ]] Tweaks to formatting of function calls.

runrevmark committed:
[[ Foundation ]] Make sure low-level scan operations check for empty needle.

The rule is that no string contains empty, therefore any low-level
operation which is equivalent to searching for a string (in some
manner) should fail if the needle is empty.

runrevmark committed:
[[ Chunks ]] Added some tests for multichar delimiters

Tests have been added to test multichar delimiters when using the
item chunk. They cover non-symmetric, symmetric and overlapping symmetric
cases (e.g. '<>', '!!' and '!!' where '!!!' is used to delimit strings).

runrevmark committed:
[[ Strings ]] Make 'begins with' and 'ends with' use specific op

This adds MCStringBeginsWithR and MCStringEndsWithR operations to
libfoundation which are the same as MCStringBeginsWith and
MCStringEndsWith except that they return the match length in the
haystack string.

These two operations are then used by EvalBeginsWith and EvalEndsWith
in the engine.

runrevmark committed:
[[ StringOptions ]] Make MCStringOptions easier to manipulate.

MCStringOptions has been reordered slightly so that fold requirement
is specified by bit 1, and the normalize requirement is specified by
bit 0.

This means that MCExecContext can store MCStringOptions directly,
and manipulate its bits when the caseSensitive and formSensitive
properties are changed.

runrevmark committed:
[[ Foundation ]] Fix MCStringUppercase for native.

The native version of uppercasing was actually lowercasing. This
has been fixed, along with lcs tests added for toUpper and toLower.

runrevmark committed:
[[ Foundation ]] Simplify BeginsWith/EndsWith APIs

The MCStringBeginsWith and MCStringEndsWith APIs now have optional
return parameters for the haystack match length. This removes the
need for 'R' variants previously introduced.

runrevmark committed:
[[ Foundation ]] Remove unnecessary 'U' modifiers.

The 'U' modifiers to the enum defining the kMCStringOptions were
unnecessary and have been removed.

runrevmark committed:
[[ Foundation ]] Remove unused foundation-string.h header.

runrevmark committed:
[[ Foundation ]] Remove misleading comment.

The top of the foundation-string.cpp file claimed MCStringRef didn't yet
work for Unicode... This has been a lie for a long time.

peter-b commented:
Reviewed 5 of 12 files at r1, 1 of 7 files at r2, 1 of 5 files at r3, 4 of 11 files at r4.
Review status: 11 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

_[libfoundation/include/foundation.h, line 1929 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P-FJFNwWhlmS3IgMO:-K3P-FJFNwWhlmS3IgMP:628271416) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/include/foundation.h#L1929)):_
If `2` vs `2U` doesn't make a difference, maybe revert these changes.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 91 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P5jwPvbPvWDXWMSYg:-K3P5jwPvbPvWDXWMSYh:-1830891449) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L91)):_
Maybe `kMCStringFoldWindows1252_Mapping` and `kMCStringFoldMacRoman_Mapping` should be explicitly `const`.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 101 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P8DsX83KUITpbnOmM:-K3P8DsX83KUITpbnOmN:1045573882) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L101)):_
I don't think this is correct.  For example, `{` (`0x7B`) is not cased, so surely `__MCNativeChar_CheckedFold()` should return false for it.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 1065 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3PCBdrIY3co5gvocvd:-K3PCBdrIY3co5gvocve:-113685301) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L1065)):_
The argument should be `char_t *x_chars` because it is mutated.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

<!-- Sent from Reviewable.io -->

peter-b commented:
Review status: 11 of 18 files reviewed at latest revision, 4 unresolved discussions.

---

_[libfoundation/include/foundation.h, line 1929 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P-FJFNwWhlmS3IgMO:-K3PDxSQ6XfX4w6qzoAl:-1133596915) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/include/foundation.h#L1929)):_
(Or modify them; obviously the bit constants are still needed!)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

<!-- Sent from Reviewable.io -->

peter-b commented:
Reviewed 1 of 11 files at r4, 7 of 7 files at r5.
Review status: 17 of 18 files reviewed at latest revision, 3 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

<!-- Sent from Reviewable.io -->

runrevmark commented:
Review status: 17 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

_[libfoundation/src/foundation-string-native.cpp.h, line 101 [r4]](https://reviewable.io:443/reviews/livecode/livecode/3267#-K3P8DsX83KUITpbnOmM:-K3PFEDwHKy2ckE46ccV:1083930167) ([raw file](https://github.com/livecode/livecode/blob/402409e08f5c395a93a8f98880b3837e92f43c03/libfoundation/src/foundation-string-native.cpp.h#L101)):_
Well, it is a performance optimization if CheckedFold returns false - we need to create the appropriate table at some point to use. I'll add a TODO comment about it.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267)_

<!-- Sent from Reviewable.io -->

runrevmark committed:
[[ Foundation ]] Minor tweaks to native string ops.

The casing tables have been marked as const.

A note has been added about future improvement to CheckedFold.

The prefix of the mutable arguments to Uppercase and Lowercase has been corrected.

peter-b commented:
This is going to be :sparkles:awesome:sparkles:

@livecode-vulcan review ok 9b18a1c

-WIP

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267#-:-K3PVChV3CQMnemDXxo4:1414285563)_

<!-- Sent from Reviewable.io -->

livecode-vulcan commented:
💙 review by @peter-b ok 9b18a1cc472fa7a11b091dc8fea94e7263c3a44f

runrevmark committed:
[[ Foundation ]] Increase prefolded needle limit to 64 bytes.

The limit at which a needle is prefolded has been put up to 64 bytes
which is the size of a standard cache line.

runrevmark committed:
[[ Bug 15855 ]] Added release note.

peter-b commented:
@livecode-vulcan review ok 1962396

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/livecode/livecode/3267#-:-K3PZnJW7ZHRT0rxx6pd:-1167224084)_

<!-- Sent from Reviewable.io -->

livecode-vulcan commented:
💙 review by @peter-b ok 19623961fd0e2f14bef0cd2d1e7fccd5b6a8ff4b

livecode-vulcan commented:
😎 test success 19623961fd0e2f14bef0cd2d1e7fccd5b6a8ff4b
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1195)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1206)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1192)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1199)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1210)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1212)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1214)
